File: app/src/main/java/cu/lt/joe/mc/lyrics/activities/ListActivity.java
package cu.lt.joe.mc.lyrics.activities;

import android.animation.Animator;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.util.DisplayMetrics;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import androidx.palette.graphics.Palette;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.google.android.material.circularreveal.CircularRevealCompat;
import java.util.Objects;
import cu.lt.joe.mc.lyrics.R;
import cu.lt.joe.mc.lyrics.databinding.ListActivityLayoutBinding;
import cu.lt.joe.mc.lyrics.utils.Utils;

public class ListActivity extends BaseActivity
{
    private ListActivityLayoutBinding binding;
    private RecyclerView list;
    private Palette palette;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.list_activity_layout);
        list = binding.list;
        list.addItemDecoration(new RecyclerView.ItemDecoration()
        {
            @Override
            public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state)
            {
                if (parent.getChildAdapterPosition(view) == 0)
                    outRect.top = Utils.dpToPx(ListActivity.this, 66);
            }
        });
        binding.setActivity(this);
    }

    public RecyclerView getList()
    {
        return list;
    }

    public void setBackground(int resourceId, float scaleFactor, float blurRadius)
    {
        Bitmap backgroundBitmap = ((BitmapDrawable) Objects.requireNonNull(ContextCompat.getDrawable(this, resourceId))).getBitmap();
        Glide.with(this).load(Utils.getBlurredBitmap(this, backgroundBitmap, scaleFactor, blurRadius)).placeholder(resourceId).centerCrop().into(binding.listBackgroundIv);
        palette = Palette.from(backgroundBitmap).generate();
    }

    public Palette getPalette()
    {
        return palette;
    }

    @SuppressWarnings("unchecked")
    public <T extends androidx.databinding.ViewDataBinding> T getMenuLayoutBinding()
    {
        if (this instanceof AlbumsActivity)
            return (T) binding.albumsMenuLayout;
        else
            return (T) binding.songsMenuLayout;
    }

    @Override
    public void onMenuExpanderClick(View menuExpanderView)
    {
        super.onMenuExpanderClick(menuExpanderView);
        binding.listMenuContainer.setBackground(Utils.getBlurredDrawableFromView(binding.getRoot()));
        DisplayMetrics dim = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(dim);
        int screenCenterX = (menuExpanderView.getLeft() + menuExpanderView.getRight()) / 2;
        int screenCenterY = (menuExpanderView.getTop() + menuExpanderView.getBottom()) / 2;
        float eStartRadius = 0;
        float eEndRadius = (float) Math.hypot(dim.widthPixels, dim.heightPixels);
        Animator expand = CircularRevealCompat.createCircularReveal(binding.listMenuContainer, screenCenterX, screenCenterY, eStartRadius, eEndRadius);
        expand.setDuration(300);
        binding.listMenuContainer.setVisibility(View.VISIBLE);
        expand.start();
    }

    @Override
    public void onBackPressed()
    {
        if (binding.listMenuContainer.getVisibility() == View.VISIBLE)
        {
            DisplayMetrics dim = new DisplayMetrics();
            getWindowManager().getDefaultDisplay().getMetrics(dim);
            int screenCenterX = (binding.menuExpanderItem.getLeft() + binding.menuExpanderItem.getRight()) / 2;
            int screenCenterY = (binding.menuExpanderItem.getTop() + binding.menuExpanderItem.getBottom()) / 2;
            float eStartRadius = (float) Math.hypot(dim.widthPixels, dim.heightPixels);
            float eEndRadius = 0;
            Animator collapse = CircularRevealCompat.createCircularReveal(binding.listMenuContainer, screenCenterX, screenCenterY, eStartRadius, eEndRadius);
            collapse.setDuration(300);
            collapse.addListener(new Animator.AnimatorListener()
            {
                @Override
                public void onAnimationStart(Animator animator)
                {
                }

                @Override
                public void onAnimationEnd(Animator animator)
                {
                    binding.listMenuContainer.setVisibility(View.GONE);
                }

                @Override
                public void onAnimationCancel(Animator animator)
                {
                }

                @Override
                public void onAnimationRepeat(Animator animator)
                {
                }
            });
            collapse.start();
        }
        else
            super.onBackPressed();
    }
}

File: app/src/main/java/cu/lt/joe/mc/lyrics/activities/ListActivity.kt
package cu.lt.joe.mc.lyrics.activities

import android.animation.Animator
import android.graphics.Rect
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.View
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.palette.graphics.Palette
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.bumptech.glide.Glide
import com.google.android.material.circularreveal.CircularRevealCompat
import cu.lt.joe.mc.lyrics.R
import cu.lt.joe.mc.lyrics.databinding.ListActivityLayoutBinding
import cu.lt.joe.mc.lyrics.utils.Utils
import java.util.Objects
import kotlin.math.hypot

open class ListActivity : BaseActivity() {
    private lateinit var binding: ListActivityLayoutBinding
    protected lateinit var list: RecyclerView
    protected lateinit var palette: Palette

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.list_activity_layout)
        list = binding.list
        list.addItemDecoration(object : ItemDecoration() {
            override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
                if (parent.getChildAdapterPosition(view) == 0) outRect.top =
                    Utils.dpToPx(this@ListActivity, 66)
            }
        })
        binding.activity = this
    }

    fun setBackground(resourceId: Int, scaleFactor: Float, blurRadius: Float) {
        val backgroundBitmap =
            (Objects.requireNonNull(ContextCompat.getDrawable(this, resourceId)) as BitmapDrawable).bitmap
        Glide.with(this)
            .load(Utils.getBlurredBitmap(this, backgroundBitmap, scaleFactor, blurRadius))
            .placeholder(resourceId).centerCrop().into(binding.listBackgroundIv)
        palette = Palette.from(backgroundBitmap).generate()
    }

    fun <T : ViewDataBinding?> getMenuLayoutBinding(): T {
        return if (this is AlbumsActivity) binding.albumsMenuLayout as T else binding.songsMenuLayout as T
    }

    override fun onMenuExpanderClick(menuExpanderView: View) {
        super.onMenuExpanderClick(menuExpanderView)
        binding.listMenuContainer.background = Utils.getBlurredDrawableFromView(binding.root)
        val dim = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(dim)
        val screenCenterX = (menuExpanderView.left + menuExpanderView.right) / 2
        val screenCenterY = (menuExpanderView.top + menuExpanderView.bottom) / 2
        val eStartRadius = 0f
        val eEndRadius =
            hypot(dim.widthPixels.toDouble(), dim.heightPixels.toDouble()).toFloat()
        val expand =
            CircularRevealCompat.createCircularReveal(binding.listMenuContainer, screenCenterX.toFloat(), screenCenterY.toFloat(), eStartRadius, eEndRadius)
        expand.duration = 300
        binding.listMenuContainer.visibility = View.VISIBLE
        expand.start()
    }

    override fun onBackPressed() {
        if (binding.listMenuContainer.visibility == View.VISIBLE) {
            val dim = DisplayMetrics()
            windowManager.defaultDisplay.getMetrics(dim)
            val screenCenterX =
                (binding.menuExpanderItem.left + binding.menuExpanderItem.right) / 2
            val screenCenterY =
                (binding.menuExpanderItem.top + binding.menuExpanderItem.bottom) / 2
            val eStartRadius =
                hypot(dim.widthPixels.toDouble(), dim.heightPixels.toDouble()).toFloat()
            val eEndRadius = 0f
            val collapse =
                CircularRevealCompat.createCircularReveal(binding.listMenuContainer, screenCenterX.toFloat(), screenCenterY.toFloat(), eStartRadius, eEndRadius)
            collapse.duration = 300
            collapse.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animator: Animator) {}
                override fun onAnimationEnd(animator: Animator) {
                    binding.listMenuContainer.visibility = View.GONE
                }

                override fun onAnimationCancel(animator: Animator) {}
                override fun onAnimationRepeat(animator: Animator) {}
            })
            collapse.start()
        } else super.onBackPressed()
    }
}